### PR TITLE
Trajectory tracking using Tracy

### PR DESCRIPTION
**Pull request by antoinealb**
This pull request integrates tracy the trajectory tracker. The code is not 100% clean yet but I heard there will be quite some modifications soon (bus enumeration and whatnot), so I think this can be handled then.

It was not tested yet on robot.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -36,6 +45,72 @@ void panic_hook(const char *reason)
     NVIC_SystemReset();
 }
 
+msg_t trajectory_thread(void *p)
```
Comment (Stapelzeiger) [inline src/main.c]: I assume this is only in main.c for testing.

File: src/main.c [hunk 2]
```diff
@@ -36,6 +45,72 @@ void panic_hook(const char *reason)
...
+            /* Perform controller iteration */
+            tracy_linear_controller(&error, &input, &output);
+
+            /* Apply speed to wheels. */
+            chprintf((BaseSequentialStream *)&SDU1 , "%d %.2f %.2f\n\r",
+                     now.s, output.tangential_velocity, output.angular_velocity);
+
+
+            /* TODO: Refactor this */
+            m1_vel_setpt = (0.5f * ROBOT_RIGHT_WHEEL_DIRECTION / ROBOT_RIGHT_MOTOR_WHEEL_RADIUS) * (output.tangential_velocity / M_PI + ROBOT_MOTOR_WHEELBASE * output.angular_velocity);
```
Comment (Stapelzeiger) [inline src/main.c]: defines should be replaced by parameter module (we have two different robots running the same code)

File: src/motor_control.h [hunk 3]
```diff
@@ -1,15 +1,46 @@
...
+
 #ifdef __cplusplus
 extern "C" {
 #endif
 
+
 extern float m1_vel_setpt;
 extern float m2_vel_setpt;
 
+struct robot_traj_s {
```
Comment (Stapelzeiger) [inline src/motor_control.h]: this should be called differential_base_traj_s

File: src/motor_control.h [hunk 4]
```diff
@@ -1,15 +1,46 @@
...
+    /** Angular speed */
+    trajectory_frame_t omega[ROBOT_TRAJ_LEN];
+
+    /** Guarantees exclusive access to the structure. */
+    mutex_t lock;
+};
+
+extern struct robot_traj_s robot_traj;
+
+extern struct robot_base_pose_2d_s robot_pose;
```
Comment (Stapelzeiger) [inline src/motor_control.h]: rename to robot_base_pose?

File: src/msg_callbacks.c [hunk 5]
```diff
@@ -49,10 +51,49 @@ void message_vel_callback(void *p, int argc, cmp_ctx_t *input)
...
+        cmp_read_int(input, &newtraj[i].date.us);
+        cmp_read_float(input, &newtraj[i].val);
+    }
+
+    /* TODO: Use smarter merging algorithm. */
+    memset(traj, 0, traj_len * sizeof(trajectory_frame_t));
+    memcpy(traj, newtraj, point_count * sizeof(trajectory_frame_t));
+}
+
+void message_traj_callback(void *p, int argc, cmp_ctx_t *input)
```
Comment (Stapelzeiger) [inline src/msg_callbacks.c]: rename base_traj

File: src/trajectories.c [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+{
+    int write_index, i;
+
+    write_index = trajectory_find_point_after(traj, traj_len, newtraj[0].date);
+
+    for (i = 0; i < newtraj_len; ++i) {
+        traj[(write_index + i) % traj_len].val = newtraj[i].val;
+        traj[(write_index + i) % traj_len].date.s = newtraj[i].date.s;
+        traj[(write_index + i) % traj_len].date.us = newtraj[i].date.us;
+    }
```
Comment (Stapelzeiger) [inline src/trajectories.c]: A check that newtraj_len < traj_len is needed.
And a related issue: we don't want to overwrite trajectory points between now and the start of the new trajectory.
I suggest we add a current time index (=read pointer for control) to trajectories.

File: src/trajectories.c [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        index ++;
+        if (index >= traj_len) {
+            /* Given date is before the earliest point in trajectory. */
+            return -1;
+        }
+    }
+
+    while (timestamp_unix_compare(traj[index % traj_len].date, date) < 0) {
+        index ++;
+        if (index >= traj_len+1) {
```
Comment (Stapelzeiger) [inline src/trajectories.c]: why is the +1 needed? (the code is not easy to understand, maybe some comments would help)

File: src/trajectories.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+#define TRAJECTORIES_H
+
+#include "unix_timestamp.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    typedef struct {
+        unix_timestamp_t date;
```
Comment (Stapelzeiger) [inline src/trajectories.h]: This needs to be discussed.
We should either allow variable sampling periods and have a timestamp, or only store one timestamp for the beginning of the trajectory and have a fixed sampling rate.

File: src/trajectories.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+        float val;
+    } trajectory_frame_t;
+
+    void trajectory_merge(trajectory_frame_t *traj, int traj_len, trajectory_frame_t *newtraj, int newtraj_len);
```
Comment (Stapelzeiger) [inline src/trajectories.h]: indentation?

**Comment by Stapelzeiger** [comment]
The pose info should not be in motor_control.c/h. I suggest moving it to robot_pose.c/h.
Also the way we handle time in trajectories should be discussed. The current implementation with one timestamp per variable per sample per actuator seems too inefficient to me (both in terms of memory usage and search/merge time) unless we would send "sparse" trajectories with a variable sampling rate.
Lastly I think trajectories need an explicit read pointer so we don't delete point that haven't been executed yet.

File: src/main.c (hunk 1 above)
Comment (antoinealb) [inline src/main.c]: Correct. Don't really know where to put it for now.

File: src/trajectories.c (hunk 6 above)
Comment (antoinealb) [inline src/trajectories.c]: We could. But this function is buggy, I will rewrite it this w-e probably.

File: src/trajectories.h (hunk 8 above)
Comment (antoinealb) [inline src/trajectories.h]: I don't see the benefits of having a start date and a sampling rate. RAM usage ?

**Comment by antoinealb** [comment]
But we need to delete point that haven't been executed, in order to do an emergency rampdown for example.

File: src/trajectories.c (hunk 6 above)
Comment (31415us) [inline src/trajectories.c]: @Stapelzeiger : could you quickly explain why newtraj_len has to be smaller than traj_len?

Comment (antoinealb) [inline src/trajectories.c]: `traj_len` is the length of the buffer (should rename it to max_traj_len) or something.

Comment (31415us) [inline src/trajectories.c]: @antoinealb : thanks (= yeah renaming might be a good idea

**Comment by Stapelzeiger** [comment]
Sure, what I meant was that a new trajectory overwrites old waypoints when wrapping around the circular bufffer end. This is generally what we want, but only if the overwritten waypoins are no longer used. (i.e. the write pointer must not overrun the read pointer)

**Comment by antoinealb** [comment]
Oh ok. I can add a `now` parameter to `trajectory_merge` to provide this information.

File: src/trajectories.h (hunk 8 above)
Comment (Stapelzeiger) [inline src/trajectories.h]: mostly time for trajectory merging, but RAM too.

**Comment by Stapelzeiger** [comment]
I would rather have a data structure for the trajectory, containing buffer pointer, buffer length, read pointer and sampling frequency.

**Comment by 31415us** [comment]
hmm does it make sense to put the sampling frequency in a structure? i think it would be best to have it as a constant in some 'config'. Do you see any advantage of changing the sampling frequency on the fly?

**Comment by Stapelzeiger** [comment]
either a variable or a config (to be discussed), but in any case to replace the timestamp per datapoint.
(I forgot in the above list, the timestamp of the read pointer needs to be stored as well)

**Comment by antoinealb** [comment]
Here is my proposal for a new trajectory module: https://gist.github.com/antoinealb/17e26cf7d0e1702097bd

Does it fix your concerns ?

**Comment by antoinealb** [comment]
I changed the trajectory module with a fixed version of my gist.

**Comment by antoinealb** [comment]
Anybody to review it ?

**Comment by antoinealb** [comment]
Can someone review and merge this ? it is getting quite big :)

**Comment by antoinealb** [comment]
@Stapelzeiger fixed your concerns:

* out of order chunks
* round to nearest point
* prevent time drift in `trajectory_read`.

**Pull request merged**